monkey/monkey_island/cc/repositories/plugin_archive_parser.py

```diff
@@ -97,3 +124,83 @@ def _safe_extract_file(tar: TarFile, filename: str) -> IO[bytes]:
     # `Optional[IO[bytes]]`.
     assert file_obj is not None
     return file_obj
+
+
+def get_plugin_source_vendors(plugin_source_tar: TarFile) -> Sequence[TarInfo]:
+    plugin_source_vendors = [
+        member
+        for member in plugin_source_tar.getmembers()
+        if (member.name in [vendor_dir.value for vendor_dir in VendorDirName] and member.isdir())
+    ]
+    return plugin_source_vendors
+
+
+def _parse_plugin_with_generic_vendor(
+    manifest: AgentPluginManifest,
+    schema: Dict[str, Any],
+    source: bytes,
+) -> Mapping[OperatingSystem, AgentPlugin]:
+    plugin = AgentPlugin(
+        plugin_manifest=manifest,
+        config_schema=schema,
+        source_archive=source,
+        host_operating_systems=(OperatingSystem.LINUX, OperatingSystem.WINDOWS),
+    )
+
+    return {OperatingSystem.LINUX: plugin, OperatingSystem.WINDOWS: plugin}
+
+
+def _parse_plugin_with_multiple_vendors(
+    plugin_source_tar: TarFile,
+    plugin_source_vendors: Sequence[TarInfo],
+    manifest: AgentPluginManifest,
+    schema: Dict[str, Any],
+) -> Mapping[OperatingSystem, AgentPlugin]:
+    os_specific_plugin_source_archives = get_os_specific_plugin_source_archives(
+        plugin_source_tar, plugin_source_vendors
+    )
+
+    parsed_plugin = {}
+    for os_, os_specific_plugin_source_archive in os_specific_plugin_source_archives.items():
+        parsed_plugin[os_] = AgentPlugin(
+            plugin_manifest=manifest,
+            config_schema=schema,
+            source_archive=os_specific_plugin_source_archive,
+            host_operating_systems=(os_,),
+        )
+
+    return parsed_plugin
+
+
+def get_os_specific_plugin_source_archives(
+    plugin_source_tar, plugin_source_vendors
+) -> Mapping[OperatingSystem, bytes]:
+
+    os_specific_plugin_source_archives = {}
+
+    for vendor in plugin_source_vendors:
+        if vendor.name == VendorDirName.LINUX_VENDOR.value:
+            vendor_os = OperatingSystem.LINUX
+        elif vendor.name == VendorDirName.WINDOWS_VENDOR.value:
+            vendor_os = OperatingSystem.WINDOWS
+        else:
+            logger.info(
+                f"Operating system of vendor directory ({vendor.name}) not recognised."
+                f"Vendor directory should be named one of the following names based on the"
+                f"supported operating systems: {[v.value for v in VendorDirName]}"
+            )
+            continue
+
+        file_handle = io.BytesIO()
+        with TarFile(fileobj=file_handle, mode="w") as os_specific_plugin_tar:
+            for item in [
+                tarinfo
+                for tarinfo in plugin_source_tar.getmembers()
+                if tarinfo.name.startswith(f"{vendor.name}/")
+                or not tarinfo.name.startswith("vendor")
```

Comment:
I'm not sure what this logic is meant to achieve, so I didn't refactor, but if it's correct we should use the vendor enum